crates/sound/src/audio_backends/pw_backend.rs

```diff
@@ -400,18 +403,46 @@ impl AudioBackend for PwBackend {
         if let Some(stream) = stream_hash.get(&stream_id) {
             stream.disconnect().expect("could not disconnect stream");
         } else {
-            error!("stream not found for the given stream_id");
+            return Err(Error::StreamWithIdNotFound(stream_id));
         };
         if let Some(stream_lis) = stream_listener.get(&stream_id) {
             //unregister stream
             std::mem::drop(stream_lis);
         } else {
-            error!("stream listener not found for the given stream_id");
+            return Err(Error::StreamListenerWithIdNotFound(stream_id));
         };
 
-        if stream_states[stream_id as usize] == VIRTIO_SND_R_PCM_PREPARE {
-            stream_states[stream_id as usize] = VIRTIO_SND_R_PCM_SET_PARAMS;
-        }
+        stream_states[stream_id as usize] = VIRTIO_SND_R_PCM_RELEASE;
+        self.thread_loop.unlock();
+        Ok(())
+    }
+
+    fn start(&self, stream_id: u32) -> Result<()> {
+        debug!("pipewire start");
+        self.thread_loop.lock();
+        let mut stream_states = self.stream_states.write().unwrap();
+        let stream_hash = self.stream_hash.read().unwrap();
+        if let Some(stream) = stream_hash.get(&stream_id) {
+            stream.set_active(true).expect("could not start stream");
+        } else {
+            return Err(Error::StreamWithIdNotFound(stream_id));
+        };
+        stream_states[stream_id as usize] = VIRTIO_SND_R_PCM_START;
+        self.thread_loop.unlock();
+        Ok(())
+    }
+
+    fn stop(&self, stream_id: u32) -> Result<()> {
+        debug!("pipewire stop");
+        self.thread_loop.lock();
+        let mut stream_states = self.stream_states.write().unwrap();
+        let stream_hash = self.stream_hash.read().unwrap();
+        if let Some(stream) = stream_hash.get(&stream_id) {
+            stream.set_active(false).expect("could not stop stream");
+        } else {
+            return Err(Error::StreamWithIdNotFound(stream_id));
+        };
```

Comment:
```suggestion
        if let Some(stream) = stream_hash.get(&stream_id) {
            stream.set_active(false).expect("could not stop stream");
            stream_states[stream_id as usize] = VIRTIO_SND_R_PCM_STOP;
            self.thread_loop.unlock();
            Ok(())
        } else {
            Err(Error::StreamWithIdNotFound(stream_id))
        }
```
I would write it like this instead